### PR TITLE
Babel plugin: debug-label

### LINKED ISSUE #487: Babel plugins: Adding debug label and fast refresh support
Continued from #388.

We want to support the following with babel plugins:
- Adding debug labels to atoms
- Fast Refresh support

### PR DESCRIPTION
**Pull request by Thisen**
This change adds a Babel plugin to Jotai for adding debug labels automatically to atoms.

Very much WIP.

Partially fixes #487, as I want to split the PRs as much as possible.

The plan is to create the following plugins:
- `jotai/babel-plugin/debug-label` - Handles debug label
- `jotai/babel-plugin/react-refresh` - Handles React refresh support

And the following preset to compose the plugins:
- `jotai/babel-preset`

### CODE REVIEW

**Comment by dai-shi** [comment]
Is this ready to play with? How can we try?

**Comment by Thisen** [comment]
> Is this ready to play with? How can we try?

If you want to try it, you can use the CSB build and add this to your `.babelrc`:

```json
{
  "presets": [],
  "plugins": ["jotai/babel-plugin/debug-label"]
}

```

**Comment by dai-shi** [comment]
https://codesandbox.io/s/jotai-nextjs-babel-plugin-kc8sd
Is this something you expected? Got "Internal Server Error"

File: package.json [hunk 1]
```diff
@@ -82,6 +82,12 @@
       "module": "./esm/urql.js",
       "import": "./esm/urql.mjs",
       "default": "./urql.js"
+    },
+    "./babel-plugin/debug-label": {
+      "types": "./babelPluginDefaultLabel.d.ts",
+      "module": "./esm/babelPluginDefaultLabel.js",
+      "import": "./esm/babelPluginDefaultLabel.mjs",
+      "default": "./babelPluginDefaultLabel.js"
```
Comment (dai-shi) [inline package.json:90]: As I tried, with nextjs on codesandbox, it doesn't seem to understand "exports". So, can we keep the file structure for old bundlers/envs?
```suggestion
    "./babel-plugin-debug-label": {
      "types": "./babel-plugin-debug-label.d.ts",
      "module": "./esm/babel-plugin-debug-label.js",
      "import": "./esm/babel-plugin-debug-label.mjs",
      "default": "./babel-plugin-debug-label.js"
```

Or, we can use `babelPluginDebugLabel` instead if we prefer camelCase.

Comment (Thisen) [reply package.json:90]: I actually wanted to restructure it to `./babel/plugin-debug-label`, because it would align with `./babel/plugin-react-refresh` and `./babel/preset`. If we want to support old bundlers for this, we will be "splitting" the `babel` packages.

Comment (dai-shi) [reply package.json:90]: `./babel/plugin-debug-label` looks nice!

Please patch rollup.confg.js:
```diff
diff --git a/rollup.config.js b/rollup.config.js
index 1a931ec..1c9e589 100644
--- a/rollup.config.js
+++ b/rollup.config.js
@@ -72,7 +72,7 @@ function createCommonJSConfig(input, output) {
 export default function (args) {
   let c = Object.keys(args).find((key) => key.startsWith('config-'))
   if (c) {
-    c = c.slice('config-'.length)
+    c = c.slice('config-'.length).replaceAll('_', '/')
     return [
       createCommonJSConfig(`src/${c}.ts`, `dist/${c}.js`),
       createESMConfig(`src/${c}.ts`, `dist/esm/${c}.mjs`),
```

and then:
```suggestion
    "./babel/plugin-debug-label": {
      "types": "./babel/plugin-debug-label.d.ts",
      "module": "./esm/babel/plugin-debuglabel.js",
      "import": "./esm/babel/plugin-debuglabel.mjs",
      "default": "./babel/plugin-debuglabel.js"
```

**Comment by dai-shi** [comment]
It works well in codesandbox. nextjs with babel config.
The screenshot with react devtools, using `<Provider>`:
unknown

**Comment by Thisen** [comment]
I think it's getting close to ready. Please review and try it.

Happy to get more test cases.

File: src/babel/plugin-debug-label.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        const { node } = nodePath
+        if (
+          t.isCallExpression(node.declaration) &&
+          isAtom(t, node.declaration.callee)
+        ) {
+          const filename = state.filename || 'unknown'
+
+          let displayName = path.basename(filename, path.extname(filename))
+
+          console.log(state)
```
Comment (Thisen) [inline src/babel/plugin-debug-label.ts:20]:
```suggestion
```

File: tests/babel/plugin-debug-label.test.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      `
+      export const countAtom = atom(0);
+      export default atom(0);
+    `,
+      'atoms/index.ts'
+    )
+  ).toMatchInlineSnapshot(`
+    "export const countAtom = atom(0);
+    countAtom.debugLabel = \\"countAtom\\";
+    const atoms = atom(0);
```
Comment (dai-shi) [inline tests/babel/plugin-debug-label.test.ts:65]: How does this resolve naming conflict?

**Review by dai-shi — APPROVED** [review]
I think this is good for the initial version.
There could be many edge cases that need to be covered later, but it's more important to let it go.
Let's label this as "experimental" in the release note.

File: tests/babel/plugin-debug-label.test.ts (hunk 3 above)
Comment (Thisen) [reply tests/babel/plugin-debug-label.test.ts:65]: Good catch. If someone has an atom called `atoms`, it would break - Good idea how to guard against it?

Comment (dai-shi) [reply tests/babel/plugin-debug-label.test.ts:65]: It can be any variable or even a function declaration `function atom() { return 1 }`.
I would simply ignore if it finds a conflict.

File: .codesandbox/ci.json [hunk 4]
```diff
@@ -5,7 +5,8 @@
     "react-typescript-react-ts",
     "simple-react-browserify-x9yni",
     "simple-snowpack-react-o1gmx",
-    "next-js-uo1h0"
+    "next-js-uo1h0",
+    "jotai-nextjs-with-babel-config-kc8sd"
```
Comment (dai-shi) [inline .codesandbox/ci.json:9]:
```suggestion
    "next-js-with-custom-babel-config-kc8sd",
    "react-with-custom-babel-config-2dpyh"
```

**Comment by dai-shi** [comment]
Yay, it works in non-container codesandbox.

image
